lib/src/Dscp/DB/SQLite/Queries.hs

```diff
@@ -155,6 +224,184 @@ getStudentTransactions student = do
+getProvenStudentTransactionsSince :: DBM m => Id Student -> UTCTime -> m [(MerkleProof PrivateTx, [(Word32, PrivateTx)])]
+getProvenStudentTransactionsSince studentId sinceTime = do
+    transaction $ do
+        -- Contains `(tx, idx, blockId)` map.
+        txsBlockList <- getTxsBlockMap
+
+        -- Bake `blockId -> [(tx, idx)]` map.
+        let txsBlockMap = groupToAssocWith (_tibBlockId, _tibTx) txsBlockList
+
+        results <- forM txsBlockMap $ \(blockId, transactions) -> do
+            blockData <- getBlockData blockId
+
+            let tree    = _bdTree blockData
+                indiced = [(idx, tx) | TxWithIdx tx idx <- transactions]
+                mapping = Map.fromList indiced
+                pruned  = fillEmptyMerkleTree mapping tree
+
+            return (pruned, indiced)
+
+        return [(proof, txs) | (Just proof, txs) <- results]
+
+  where
+    groupToAssocWith :: Ord k => Ord k => (a -> k, a -> v) -> [a] -> [(k, [v])]
+    groupToAssocWith (key, value) =
+        Map.toList . foldr' push Map.empty
+      where
+        push a = Map.insertWith (++) (key a) [value a]
+
+    -- Returns, effectively, `[(tx, idx, blockId)]`.
+    getTxsBlockMap :: DBM m => m [TxInBlock]
+    getTxsBlockMap = do
+        query getTxsBlockMapQuery (studentId, sinceTime)
+      where
+        getTxsBlockMapQuery = [q|
+            -- getTxsBlockMapQuery
+
+            select     Submissions.student_addr,
+                       Submissions.contents_hash,
+                       Assignments.course_id,
+                       Assignments.contents_hash,
+                       Assignments.type,
+                       Assignments.desc,
+                       Submissions.signature,
+                       Transactions.grade,
+                       Transactions.time,
+                       Transactions.idx,
+                       BlockTxs.blk_idx
+
+            from       BlockTxs
+            left join  Transactions
+                   on  tx_hash = Transactions.hash
+
+            left join  Submissions
+                   on  Transactions.submission_hash = Submissions.hash
+
+            left join  StudentAssignments
+                   on  StudentAssignments.assignment_hash = Submissions.assignment_hash
+
+            left join  Assignments
+                   on  StudentAssignments.assignment_hash = Assignments.hash
+
+            where      StudentAssignments.student_addr  =  ?
+                  and  time                            >=  ?
+                  and  Transactions.idx                <> -1
+        |]
+
+    -- Returns `PrivateBlock` in normalized format, with metadata.
+    getBlockData :: DBM m => Word32 -> m BlockData
+    getBlockData blockIdx = do
+        (listToMaybe <$> query getBlockDataQuery (Only blockIdx))
+            `assertJust` BlockWithIndexDoesNotExist blockIdx
+      where
+        -- TODO: prune fields that aren't needed.
```

Comment:
Is it still relevant?

Reply:
Nope. I choose to hold all data in th end.